### PR TITLE
FM2-476: Add support for searching Medication Requests based on status

### PR DESCRIPTION
**Pull request by mherman22**
## Description of what I changed

Updated the following in order to support searching for MedicationRequests based on status(https://www.hl7.org/fhir/medicationrequest-definitions.html#MedicationRequest.status);-

- MedicationRequestFhirResourceProvider(s).
- Service.
-  Dao.
- unit tests.

## Issue I worked on

see https://issues.openmrs.org/browse/FM2-476

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by mherman22** [comment]
Still a WIP

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirMedicationRequestDaoImpl.java [hunk 1]
```diff
@@ -47,13 +53,33 @@ protected void setupSearchParams(Criteria criteria, SearchParameterMap theParams
...
 			}
 		});
 	}
 	
+	private Optional<Criterion> handleMedicationRequestStatus(TokenAndListParam tokenAndListParam) {
+		return handleAndListParam(tokenAndListParam, token -> {
+			if (token.getValue() != null) {
+				try {
+					return Optional.of(
+					    eq("status", MedicationRequest.MedicationRequestStatus.valueOf(token.getValue().toUpperCase())));
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirMedicationRequestDaoImpl.java:72]: Is there a `status` field on the OpenMRS model for `DrugOrder`?

**Review by ibacher — COMMENTED** [review]
So, first @mherman22 thanks for the effort here.

There are a few problems:

1. There is no `status` field in the OpenMRS data model. In 2.2+ there is a `fulfilerStatus`. So, at the very least, while we should add a search parameter for `MedicationRequest`, the implementation needs to go in a folder targetted at the 2.2+ release of OpenMRS.
2. It's not clear that `fulfillerStatus` really maps to the FHIR MedicaitonRequest `status` field. The only value they have in common is `completed`. This implies that the `status` we need to search on needs to be generated from more data than just a single column search.
3. Ultimately, I don't think this ticket should've been marked "Ready for Work" because it really isn't given 2 above.

For that reason, I'm going to close this PR and move the ticket to "Needs Design."

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirMedicationRequestDaoImpl.java (hunk 1 above)
Comment (mherman22) [reply api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirMedicationRequestDaoImpl.java:72]: > Is there a `status` field on the OpenMRS model for `DrugOrder`?

not really , drug order doesn't seem to have a status anywhere on the [data model](https://docs.openmrs.org/datamodel/)

**Comment by mherman22** [comment]
> So, first @mherman22 thanks for the effort here.
> 
> There are a few problems:
> 
> 1. There is no `status` field in the OpenMRS data model. In 2.2+ there is a `fulfilerStatus`. So, at the very least, while we should add a search parameter for `MedicationRequest`, the implementation needs to go in a folder targetted at the 2.2+ release of OpenMRS.
> 2. It's not clear that `fulfillerStatus` really maps to the FHIR MedicaitonRequest `status` field. The only value they have in common is `completed`. This implies that the `status` we need to search on needs to be generated from more data than just a single column search.
> 3. Ultimately, I don't think this ticket should've been marked "Ready for Work" because it really isn't given 2 above.
> 
> For that reason, I'm going to close this PR and move the ticket to "Needs Design."

- After looking through the openmrs model, it is now clear to me that the Drug order doesn't have a status as you have stated above.
- However i beg to be corrected if am wrong @ibacher , according to the [medicationRequestStatusTranslator implementation class](https://github.com/openmrs/openmrs-module-fhir2/blob/master/api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/MedicationRequestStatusTranslatorImpl.java), in the [MedicationRequest.MedicationRequestStatus toFhirResource](https://github.com/openmrs/openmrs-module-fhir2/blob/f20accf25fc497b518c9d22ca4027fe9a773d241/api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/MedicationRequestStatusTranslatorImpl.java#L23) function, i thought the whole point there is that;-

1. if the Drug Order is **active** then it can be translated as **ACTIVE** in MedicationRequest.MedicationRequestStatus.
2. if it **isDiscontinuedRightNow** then it can be translated as **CANCELLED** to MedicationRequest.MedicationRequestStatus.
3. if it is null then subsquently it can be translated as NULL.

**Comment by ibacher** [comment]
@mherman22 Yep, that's the logic I was looking for! So, can you get the search to work with the same criteria? Also, we'll need some tests that actually prove that the search works.

**Comment by mherman22** [comment]
> @mherman22 Yep, that's the logic I was looking for! So, can you get the search to work with the same criteria? Also, we'll need some tests that actually prove that the search works. PS, I wouldn't worry about searching for `null` as that's not a meaningful search, just if the value is `CANCELLED` it will match requests that are cancelled (according to the translator) and if the value is `ACTIVE` if will match requests that are active (according to the translator).

jumping into that work now, thanks for the clarity

**Comment by mherman22** [comment]
Paving way for another contributor to bring this home

**Pull request closed without merging**